File: src/main/java/be/codecraft/webshop/controller/service/AddressService.java
package be.codecraft.webshop.controller.service;

import be.codecraft.webshop.controller.securityJWT.model.User;
import be.codecraft.webshop.controller.securityJWT.model.UserRepository;
import be.codecraft.webshop.datamodel.model.Address;
import be.codecraft.webshop.datamodel.model.AddressDetails;
import be.codecraft.webshop.datamodel.model.dto.AddressDTO;
import be.codecraft.webshop.datamodel.repository.AddressRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
public class AddressService {

    private final AddressRepository addressRepository;
    private final UserRepository userRepository;
    private final EntityMapper entityMapper;

    public AddressService(AddressRepository addressRepository, UserRepository userRepository, EntityMapper entityMapper) {
        this.addressRepository = addressRepository;
        this.userRepository = userRepository;
        this.entityMapper = entityMapper;
    }

    @Transactional
    public AddressDTO getAddressById(UUID addressId) {
        Address address = addressRepository.findById(addressId)
                .orElseThrow(() -> new IllegalArgumentException("Address not found for ID: " + addressId));
        return entityMapper.convertAddressToDTO(address);
    }

    @Transactional
    public AddressDTO getAddressByUserEmail(String userEmail) {
        Address address = addressRepository.findByUserEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("Address not found for user: " + userEmail));
        return entityMapper.convertAddressToDTO(address);
    }

    @Transactional
    public AddressDTO createAddress(AddressDTO addressDTO) {
        Address newAddress = entityMapper.convertAddressToEntity(addressDTO);

        Address savedAddress = addressRepository.save(newAddress);
        return entityMapper.convertAddressToDTO(savedAddress);
    }

    @Transactional
    public AddressDTO updateAddress(UUID addressId, AddressDTO addressDTO) {
        Address address = addressRepository.findById(addressId)
                .orElseThrow(() -> new IllegalArgumentException("Address not found for ID: " + addressId));

        AddressDetails addressDetails = address.getAddressDetails();
        addressDetails.setStreet(addressDTO.getStreet());
        addressDetails.setCity(addressDTO.getCity());
        addressDetails.setState(addressDTO.getState());
        addressDetails.setPostalCode(addressDTO.getPostalCode());
        addressDetails.setCountry(addressDTO.getCountry());

        addressRepository.save(address);

        return entityMapper.convertAddressToDTO(address);
    }

    @Transactional
    public void deleteAddress(UUID addressId) {
        Address address = addressRepository.findById(addressId)
                .orElseThrow(() -> new IllegalArgumentException("Address not found for ID: " + addressId));
        addressRepository.delete(address);
    }

    @Transactional
    public void deleteAddressByUserEmail(String userEmail) {
        Address address = addressRepository.findByUserEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("Address not found for user: " + userEmail));
        addressRepository.delete(address);
    }

}

File: frontend/src/app/components/shop/services/address.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Address } from '../../types';


@Injectable({
  providedIn: 'root',
})
export class AddressService {
  private apiUrl = 'http://localhost:8080/api/addresses';

  constructor(private http: HttpClient) {}


  getAddressById(addressId: string): Observable<Address> {
    return this.http.get<Address>(`${this.apiUrl}/${addressId}`);
  }

  getAddressByUserEmail(userEmail: string): Observable<Address> {
    return this.http.get<Address>(`${this.apiUrl}/user/${userEmail}`);
  }

  createAddress(address: Address): Observable<Address> {
    return this.http.post<Address>(this.apiUrl, address);
  }

  updateAddress(addressId: string, address: Address): Observable<Address> {
    return this.http.put<Address>(`${this.apiUrl}/${addressId}`, address);
  }

  deleteAddress(addressId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${addressId}`);
  }

  deleteAddressByUserEmail(userEmail: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/user/${userEmail}`);
  }
}
